### PR TITLE
Bridges of LI.FI across chains, the contract is LiFiDiamond_v2

### PR DESCRIPTION
**Pull request by lequangphu**
## Thank you for contributing to Spellbook 🪄
Please open the PR in **draft** and mark as ready when you want to request a review. 

### Description:

Extract bridge data from event `LiFiDiamond_v2_evt_LiFiTransferStarted` on each chain, then union across chains.

---
quick links for more information:
- [README.md](https://github.com/duneanalytics/spellbook/blob/main/README.md)
- [spellbook docs](https://github.com/duneanalytics/spellbook/tree/main/docs)
- [CONTRIBUTING.md](https://github.com/duneanalytics/spellbook/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by jeff-dude** [comment]
before we dig into the failing test, i think we need to redesign this pipeline a bit.
- flip your materialization strategy approach -- i.e. materialize each chain incrementally, then have the crosschain be a view on top
- looks like the code is the same per chain -- in this case, i'd use a dbt macro for the logic and call the macro for each chain while passing in the `blockchain` parameter to differentiate. there are tons of examples in spellbook to use as reference, one to search could be the uniswap models in `dex` subproject (but plenty others too!)
- when joining from decoded to base table to get additional columns, lets use [this](https://github.com/duneanalytics/spellbook/blob/main/dbt_macros/shared/add_tx_columns.sql) shared macro. an example to call the macro is [here](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/daily_spellbook/models/_projects/gmx/event/avalanche_c/gmx_v2_avalanche_c_position_increase.sql#L289-L295)
- when using the if is incremental logic, let's use the latest macro approach like [here](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/macros/models/_decoding/uniswap_v2/uniswap_v2_trades.sql#L25-L27) rather than hardcode last 7 days
- for incremental chain models, i would expect config to contain all of [this](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/models/trades/bnb/platforms/uniswap_v2_bnb_base_trades.sql#L1-L11)
- looks like you removed the surrogate key logic, i'd put that back in and use as unique key. then update the schema yml to run unique test on this column, not the combination of columns that make up the surrogate key


let's leave it at that for now, and revisit once all this in place 🙏

**Comment by lequangphu** [comment]
> before we dig into the failing test, i think we need to redesign this pipeline a bit.
> 
> * flip your materialization strategy approach -- i.e. materialize each chain incrementally, then have the crosschain be a view on top
> * looks like the code is the same per chain -- in this case, i'd use a dbt macro for the logic and call the macro for each chain while passing in the `blockchain` parameter to differentiate. there are tons of examples in spellbook to use as reference, one to search could be the uniswap models in `dex` subproject (but plenty others too!)
> * when joining from decoded to base table to get additional columns, lets use [this](https://github.com/duneanalytics/spellbook/blob/main/dbt_macros/shared/add_tx_columns.sql) shared macro. an example to call the macro is [here](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/daily_spellbook/models/_projects/gmx/event/avalanche_c/gmx_v2_avalanche_c_position_increase.sql#L289-L295)
> * when using the if is incremental logic, let's use the latest macro approach like [here](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/macros/models/_decoding/uniswap_v2/uniswap_v2_trades.sql#L25-L27) rather than hardcode last 7 days
> * for incremental chain models, i would expect config to contain all of [this](https://github.com/duneanalytics/spellbook/blob/main/dbt_subprojects/dex/models/trades/bnb/platforms/uniswap_v2_bnb_base_trades.sql#L1-L11)
> * looks like you removed the surrogate key logic, i'd put that back in and use as unique key. then update the schema yml to run unique test on this column, not the combination of columns that make up the surrogate key
> 
> let's leave it at that for now, and revisit once all this in place 🙏

thanks, i will update the pr based on your recommendation.

**Comment by jeff-dude** [comment]
> thanks, i will update the pr based on your recommendation.

great, just tag me again when you are ready for next iteration of review 🙏

**Comment by lequangphu** [comment]
> > thanks, i will update the pr based on your recommendation.
> 
> great, just tag me again when you are ready for next iteration of review 🙏

it's green check, so i guess i'm ready.

File: dbt_subprojects/daily_spellbook/macros/project/lifi/lifi_extract_bridge_data_macro.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/macros/project/lifi/lifi_extract_bridge_data_macro.sql:20]:
```suggestion
    cast(date_trunc('day', evt_block_time) as date) as block_date,
```

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/LiFiDiamond_v2_evt_LiFiTransferStarted.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/LiFiDiamond_v2_evt_LiFiTransferStarted.sql:4]:
```suggestion
        alias = 'transfers',
```

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/LiFiDiamond_v2_evt_LiFiTransferStarted.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/LiFiDiamond_v2_evt_LiFiTransferStarted.sql:1]: dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/schema.yml
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/schema.yml:67]:
```suggestion
  - name: lifi_transfers
```

**Review by jeff-dude — CHANGES_REQUESTED** [review]
really clean code here, much appreciated!

left a few minor final changes below, please do implement when ready.

last thing: have you queried the data output yet, from CI pipeline?
https://github.com/duneanalytics/spellbook/actions/runs/12054789316/job/33613636277?pr=7172#step:12:32
you can query these tables on dune and ensure it looks as you expect, prior to merge 🙏

**Comment by lequangphu** [comment]
> really clean code here, much appreciated!
> 
> left a few minor final changes below, please do implement when ready.
> 
> last thing: have you queried the data output yet, from CI pipeline? https://github.com/duneanalytics/spellbook/actions/runs/12054789316/job/33613636277?pr=7172#step:12:32 you can query these tables on dune and ensure it looks as you expect, prior to merge 🙏

I applied your suggestion.

I want to add amount_usd to the output as what I do from line 223-427 in this query: https://dune.com/queries/4182530 . What is your recommendation of doing this in the models?

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql:6]:
```suggestion
        materialized = 'view',
        post_hook='{{ expose_spells(\'[
                                        "arbitrum"
                                        , "avalanche_c"
                                        , "bnb"
                                        , "ethereum"
                                        , "fantom"
                                        , "gnosis"
                                        , "zksync"
                                    ]\',
                                    "project",
                                    "lifi",
                                    \'["lequangphu"]\') }}'
```

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql:6]: if you want to display as a contributor, you need to have within a post-hook which will display this final crosschain view in the data explorer and then your name at the bottom

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Comment by jeff-dude** [comment]
> How would you apply them in the models?

for pricing, i would handle the way you are now. CTE which reads raw decoded data and pass into macro, then pass that CTE into new one with join to prices. i was thinking you could use a macro for the join to prices and have each model use it, but if your manual mapping of token addresses differs per chain, it may be more complex.

to summarize:
- first CTE reads only decoded data ✅ 
- second CTE (todo) maps addresses as needed per chain
- third CTE joins to prices ✅ 
- final macro call to add columns

**Comment by lequangphu** [comment]
> * first CTE reads only decoded data ✅
> * second CTE (todo) maps addresses as needed per chain
> * third CTE joins to prices ✅
> * final macro call to add columns

Done, please review!

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql:37]:
```suggestion
        and p.minute = date_trunc('minute', tokens_mapped.block_time)
        {% if is_incremental() %}
        and {{ incremental_predicate('p.minute') }}
        {% endif %}
```

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql:35]: i noticed you are forced to cast to varchar here to get join condition to work, due to extracting field from varchar json field in macro. i think you will benefit from casting the fields out of the json to your desired data type, for instance, varbinary for the address field(s) such as `sendingAssetId`

Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql:35]: it's likely worth while to cast, maybe best in the macro itself to apply to all models? then you can remove the cast in join conditions?

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql
Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql:37]: we should apply this to all chain models

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql
Comment (lequangphu) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql]: noted, thanks!

**Comment by lequangphu** [comment]
@jeff-dude the test fails because decoded table notfound, which is weird.

Screenshot 2024-12-04 at 10 08 25

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql
Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql:6]: please revisit this note here :)

File: dbt_subprojects/daily_spellbook/macros/project/lifi/lifi_extract_bridge_data_macro.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/macros/project/lifi/lifi_extract_bridge_data_macro.sql:32]: somewhere along the way, upper/lower case matters for source tables. looks like your latest commit changed this to all lowercase, while the source yml file defined in the PR uses camelcase. your best bet is to put this back to camelcase to match source file (or other way around).

tbh i haven't confirmed which step casing matters, but i'm fairly certain it's this. i don't think the dunesql engine cares, if you run on the app with either i believe it works.

**Review by jeff-dude — CHANGES_REQUESTED** [review]

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql
Comment (jeff-dude) [inline dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql:24]: i'm fairly certain your latest failure is around these addresses having single quotes. can you remove, then should default to varbinary

Comment (lequangphu) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/arbitrum/lifi_arbitrum_transfers.sql]: you catched it while i slept after pushing the changes, thanks!

File: dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql
Comment (jeff-dude) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql:6]: @lequangphu i can't merge until this one is resolved. we don't run any actions in the model config in CI env, only once merged to prod. we should ensure we have the `post_hook` setup properly like my proposal above 🙏

Comment (lequangphu) [reply dbt_subprojects/daily_spellbook/models/_projects/lifi/lifi_transfers.sql:6]: oh, i miss it, will do.

**Pull request merged**
